File: patterns/structural/adapter.js
/**
 * Adapter Pattern - JavaScript
 * 
 * Allows objects with incompatible interfaces to collaborate.
 * Acts as a bridge between two incompatible interfaces.
 * 
 * Use cases:
 * - Integrating third-party libraries
 * - Legacy code integration
 * - Interface compatibility
 */

// Target interface (what client expects)
class PaymentProcessor {
    processPayment(amount) {
        throw new Error('Method must be implemented');
    }
}

// Adaptee (existing incompatible interface)
class LegacyPaymentSystem {
    makePayment(dollars, cents) {
        return `Legacy payment: $${dollars}.${cents.toString().padStart(2, '0')}`;
    }
}

// Adapter
class LegacyPaymentAdapter extends PaymentProcessor {
    constructor(legacySystem) {
        super();
        this.legacySystem = legacySystem;
    }
    
    processPayment(amount) {
        // Convert amount to dollars and cents
        const dollars = Math.floor(amount);
        const cents = Math.round((amount - dollars) * 100);
        return this.legacySystem.makePayment(dollars, cents);
    }
}

// Another adaptee
class ThirdPartyPaymentAPI {
    pay(amountInCents) {
        return `Third-party payment: ${amountInCents} cents`;
    }
}

// Another adapter
class ThirdPartyPaymentAdapter extends PaymentProcessor {
    constructor(thirdPartyAPI) {
        super();
        this.api = thirdPartyAPI;
    }
    
    processPayment(amount) {
        const amountInCents = Math.round(amount * 100);
        return this.api.pay(amountInCents);
    }
}

// Client code
class PaymentService {
    constructor(processor) {
        this.processor = processor;
    }
    
    handlePayment(amount) {
        return this.processor.processPayment(amount);
    }
}

// Example usage
console.log('=== Adapter Pattern ===\n');

const legacySystem = new LegacyPaymentSystem();
const legacyAdapter = new LegacyPaymentAdapter(legacySystem);
const paymentService1 = new PaymentService(legacyAdapter);
console.log(paymentService1.handlePayment(99.99));

const thirdPartyAPI = new ThirdPartyPaymentAPI();
const thirdPartyAdapter = new ThirdPartyPaymentAdapter(thirdPartyAPI);
const paymentService2 = new PaymentService(thirdPartyAdapter);
console.log(paymentService2.handlePayment(99.99));

module.exports = { PaymentProcessor, LegacyPaymentAdapter, ThirdPartyPaymentAdapter, PaymentService };

File: patterns/structural/adapter.py
"""
Adapter Pattern - Python

Allows objects with incompatible interfaces to collaborate.
Acts as a bridge between two incompatible interfaces.

Use cases:
- Integrating third-party libraries
- Legacy code integration
- Interface compatibility
"""

from abc import ABC, abstractmethod


# Target interface (what client expects)
class PaymentProcessor(ABC):
    @abstractmethod
    def process_payment(self, amount):
        pass


# Adaptee (existing incompatible interface)
class LegacyPaymentSystem:
    def make_payment(self, dollars, cents):
        return f"Legacy payment: ${dollars}.{cents:02d}"


# Adapter
class LegacyPaymentAdapter(PaymentProcessor):
    def __init__(self, legacy_system):
        self.legacy_system = legacy_system
    
    def process_payment(self, amount):
        # Convert amount to dollars and cents
        dollars = int(amount)
        cents = int((amount - dollars) * 100)
        return self.legacy_system.make_payment(dollars, cents)


# Another adaptee
class ThirdPartyPaymentAPI:
    def pay(self, amount_in_cents):
        return f"Third-party payment: {amount_in_cents} cents"


# Another adapter
class ThirdPartyPaymentAdapter(PaymentProcessor):
    def __init__(self, third_party_api):
        self.api = third_party_api
    
    def process_payment(self, amount):
        amount_in_cents = int(amount * 100)
        return self.api.pay(amount_in_cents)


# Client code
class PaymentService:
    def __init__(self, processor):
        self.processor = processor
    
    def handle_payment(self, amount):
        return self.processor.process_payment(amount)


# Example usage
if __name__ == "__main__":
    print("=== Adapter Pattern ===\n")
    
    legacy_system = LegacyPaymentSystem()
    legacy_adapter = LegacyPaymentAdapter(legacy_system)
    payment_service1 = PaymentService(legacy_adapter)
    print(payment_service1.handle_payment(99.99))
    
    third_party_api = ThirdPartyPaymentAPI()
    third_party_adapter = ThirdPartyPaymentAdapter(third_party_api)
    payment_service2 = PaymentService(third_party_adapter)
    print(payment_service2.handle_payment(99.99))

